DOC-13170 Product Change- PR #143536 - metric: add /metrics endpoint with static labels (#19823)
cockroachdb/docs: src/current/v25.3/prometheus-endpoint.md

---
title: Prometheus Endpoint
summary: Export granular time-series metrics in Prometheus format to monitor a cluster's health and performance.
toc: true
---

Each node in a CockroachDB cluster exports granular time-series metrics to two available Prometheus-compatible endpoints:

- [`http://<host>:<http-port>/_status/vars`](#_status-vars)

> - {% include_cached new-in.html version="v25.3" %}[`http://<host>:<http-port>/metrics`](#metrics): an enhanced endpoint with additional static labels

Reviewer: do we note anywhere that the names/labels on this endpoint are currently subject to change? I think they'll be locked down a bit in 25.4

Author: In the `metrics` section, there is a note: "This feature is in [preview](https://www.cockroachlabs.com/docs/v25.3/cockroachdb-feature-availability) and subject to change."

Unlabeled metrics from the `_status/vars` endpoint | Labeled metrics from the `metrics` endpoint
-----------------------------------------------|-----------------------------------------
`sql_insert_count` | `sql_count{query_type="insert"}`
`sql_select_count` | `sql_count{query_type="select"}`
`sql_update_count` | `sql_count{query_type="update"}`
`sql_delete_count` | `sql_count{query_type="delete"}`

At metrics query time, labels provide a smoother user experience:

Unlabeled sum query from the `_status/vars` endpoint | Labeled sum query from the `metrics` endpoint
-----------------------------------------------|-------------------------------
`sum(sql_insert_count, sql_delete_count, sql_select_count)` | `sum(sql_count)`
This query must be modified if new types are added because they will have new metric names. | This query is resilient to new type additions.
Related metrics can be found via autocomplete in a third-party tool, but it may be unclear. | All label values can be found through a third-party query engine and used to easily construct a graph with individual lines for each label value.

> Another common scenario occurs when each label value represents a disjoint set of categories. An example here is the various certificate expiration metrics, which differ only by the specific certificate they refer to. Operators are unlikely to aggregate these, but may still want to view all certificate expiration metrics on a dashboard.
> 
> For example, the output from the `metrics` endpoint will be similar to the following:

Reviewer 2: ```suggestion
In other cases, label values can represent distinct categories not meant to be aggregated. For example, certificate expiration metrics differ only by the specific certificate type they refer to. Operators are unlikely to sum or average these, but may still want to display them side by side on a dashboard for visibility.

In this case, a single metric name like `security_certificate_expiration` is reused, with the certificate type expressed as a label. The output from the `metrics` endpoint will be similar to the following:
```

~~~
# HELP security_certificate_expiration Expiration for the CA certificate
# TYPE security_certificate_expiration gauge
security_certificate_expiration{node_id="1",tenant="demoapp",certificate_type="ca"} 1.998766953e+09
security_certificate_expiration{node_id="1",tenant="demoapp",certificate_type="ca-client-tenant"} 0
security_certificate_expiration{node_id="1",tenant="demoapp",certificate_type="node-client"} 0
security_certificate_expiration{node_id="1",tenant="demoapp",certificate_type="client-tenant"} 0
security_certificate_expiration{node_id="1",tenant="demoapp",certificate_type="ui"} 0
security_certificate_expiration{node_id="1",tenant="demoapp",certificate_type="client"} 1.840654953e+09
security_certificate_expiration{node_id="1",tenant="demoapp",certificate_type="client-ca"} 0
security_certificate_expiration{node_id="1",tenant="demoapp",certificate_type="ui-ca"} 0
security_certificate_expiration{node_id="1",tenant="demoapp",certificate_type="node"} 1.840654953e+09

> ~~~

Reviewer 2: ```suggestion
~~~

This approach avoids a proliferation of metric names while allowing third-party tools to display each certificate's expiration as a separate line in a unified graph or table.
```

{{site.data.alerts.callout_info}}
In addition to using the exported time-series data to monitor a cluster through an external system, you can write alerting rules to ensure prompt notification of critical events or issues that may require intervention or investigation. Refer to [Essential Alerts]({% link {{ page.version.version }}/essential-alerts-self-hosted.md %}) for more details.
{{site.data.alerts.end}}

> If you rely on external tools for storing and visualizing your cluster's time-series metrics, Cockroach Labs recommends that you [disable the DB Console's storage of time-series metrics]({% link {{ page.version.version }}/operational-faqs.md %}#disable-time-series-storage).

Reviewer 3: Thinking about this more, I think we should remove this recommendation. If we disable TSDB, users wouldn't be able to generate Tsdump, which is critical for escalations. We should at least call this out in the doc.

Author: removed recommendation, added sentence about tsdump.

---
title: Prometheus Endpoint
summary: Export granular time-series metrics in Prometheus format to monitor a cluster's health and performance.
toc: true
---

Each node in a CockroachDB cluster exports granular time-series metrics to two available Prometheus-compatible endpoints:

- [`http://<host>:<http-port>/_status/vars`](#_status-vars)

> - {% include_cached new-in.html version="v25.3" %}[`http://<host>:<http-port>/metrics`](#metrics): an enhanced endpoint with additional static labels

Reviewer 2: ```suggestion
- {% include_cached new-in.html version="v25.3" %}[`http://<host>:<http-port>/metrics`](#metrics): an enhanced endpoint with additional [static labels](#static-labels)
```

Author: updated

~~~
# HELP sys_cgocalls Total number of cgo calls
# TYPE sys_cgocalls counter
sys_cgocalls{node_id="1",tenant="demoapp"} 13737
# HELP sys_cpu_sys_percent Current system cpu percentage consumed by the CRDB process
# TYPE sys_cpu_sys_percent gauge
sys_cpu_sys_percent{node_id="1",tenant="demoapp"} 0.0021986027879282717
...
# HELP sql_count Number of SQL INSERT statements successfully executed (internal queries)
# TYPE sql_count counter
sql_count{node_id="1",tenant="demoapp",query_type="insert",query_internal="true"} 1281
sql_count{node_id="1",tenant="demoapp",query_type="delete"} 0
sql_count{node_id="1",tenant="demoapp",query_type="update"} 0
sql_count{node_id="1",tenant="demoapp",query_type="select",query_internal="true"} 2280
sql_count{node_id="1",tenant="demoapp",query_type="select"} 9
sql_count{node_id="1",tenant="demoapp",query_type="insert"} 15
sql_count{node_id="1",tenant="demoapp",query_type="update",query_internal="true"} 2102
sql_count{node_id="1",tenant="demoapp",query_type="delete",query_internal="true"} 1067
...
~~~

### Static labels

> One common use of static labels is to allow segmentation of a metric across various facets for later querying and aggregation. For example, rather than emitting separate metrics for `INSERT`, `SELECT`, `UPDATE`, and `DELETE` statements, the single metric `sql_count` uses the `query_type` label to distinguish between these operations. This enables operators to easily aggregate across query types (e.g., summing all SQL operations) or filter for a specific type using the appropriate value for the `query_type` label.

Reviewer 2: ```suggestion
One common use of static labels is to allow segmentation of a metric across various facets for later querying and aggregation. For example, rather than emitting separate metrics for `INSERT`, `SELECT`, `UPDATE`, and `DELETE` statements, the single metric `sql_count` uses the `query_type` label to distinguish between these operations. This enables operators to easily aggregate across query types (e.g., summing a metric for all SQL operations) or filter for a specific type using the appropriate value for the `query_type` label.
```

Author: updated
